File: CPP/ConcatenatedWords.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

class Solution {
public:
  std::vector<std::string>
  findAllConcatenatedWordsInADict(std::vector<std::string> &words) {
    std::unordered_set<std::string> set(words.begin(), words.end());
    std::vector<std::string> res;
    for (auto &w : words) {
      if (!w.empty() && canForm(w, set))
        res.push_back(w);
    }
    return res;
  }

private:
  bool canForm(const std::string &word,
               const std::unordered_set<std::string> &set) {
    int n = (int)word.size();
    std::vector<bool> dp(n + 1, false);
    dp[0] = true;
    for (int i = 1; i <= n; ++i) {
      for (int j = 0; j < i; ++j) {
        if (!dp[j] || (j == 0 && i == n))
          continue;
        if (set.count(word.substr(j, i - j))) {
          dp[i] = true;
          break;
        }
      }
    }
    return dp[n];
  }
};

void print(const std::vector<std::string> &v) {
  std::cout << "[";
  for (size_t i = 0; i < v.size(); ++i) {
    if (i)
      std::cout << ", ";
    std::cout << v[i];
  }
  std::cout << "]\n";
}

int main() {
  Solution s;
  std::vector<std::string> a = {"cat", "cats",        "catsdogcats",
                                "dog", "dogcatsdog",  "hippopotamuses",
                                "rat", "ratcatdogcat"};
  std::vector<std::string> b = {"cat", "dog", "catdog"};
  std::vector<std::string> c = {"a", "b", "ab", "abc"};
  print(s.findAllConcatenatedWordsInADict(
      a)); // [catsdogcats, dogcatsdog, ratcatdogcat]
  print(s.findAllConcatenatedWordsInADict(b)); // [catdog]
  print(s.findAllConcatenatedWordsInADict(c)); // [ab]
  return 0;
}

File: Python/ConcatenatedWords.py
from typing import List


class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        # A word qualifies if it splits into >= 2 words from the set. Word-break DP per word:
        # dp[i] = word[:i] is formable from set words. Forbidding the full-word match (j==0 and
        # i==n) guarantees every piece is strictly shorter, so reaching dp[n] needs >= 2 pieces.
        word_set = set(words)

        def can_form(word: str) -> bool:
            n = len(word)
            dp = [False] * (n + 1)
            dp[0] = True
            for i in range(1, n + 1):
                for j in range(i):
                    if not dp[j] or (j == 0 and i == n):
                        continue
                    if word[j:i] in word_set:
                        dp[i] = True
                        break
            return dp[n]

        return [w for w in words if w and can_form(w)]


if __name__ == "__main__":
    s = Solution()
    print(
        s.findAllConcatenatedWordsInADict(
            [
                "cat",
                "cats",
                "catsdogcats",
                "dog",
                "dogcatsdog",
                "hippopotamuses",
                "rat",
                "ratcatdogcat",
            ]
        )
    )
    # ['catsdogcats', 'dogcatsdog', 'ratcatdogcat']
    print(s.findAllConcatenatedWordsInADict(["cat", "dog", "catdog"]))  # ['catdog']
    print(s.findAllConcatenatedWordsInADict(["a", "b", "ab", "abc"]))  # ['ab']
